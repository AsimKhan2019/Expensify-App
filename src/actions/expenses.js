import uuid from 'uuid';
import { database } from '../firebase/firebase';

export const addExpense = expense => ({
  type: 'ADD_EXPENSE',
  expense
});

export const startAddExpense = (expenseData = {}) => {
  return dispatch => {
    const {
      description = '',
      note = '',
      amount = 0,
      createdAt = Date.now()
    } = expenseData;

    const expense = { description, note, amount, createdAt };
    return database
      .ref('/expenses')
      .push(expense)
      .then(ref => dispatch(addExpense({ id: ref.key, ...expense })));
  };
};

export const editExpense = (id, updates) => ({
  type: 'EDIT_EXPENSE',
  expense: { ...updates, id }
});

export const removeExpense = ({ id = '' } = {}) => ({
  type: 'REMOVE_EXPENSE',
  expense: { id }
});

export const setExpenses = expenses => ({
  type: 'SET_EXPENSE',
  expenses
});

export const startSetExpenses = () => {
  return dispatch => {
    return database
      .ref('/expenses')
      .once('value')
      .then(snapshot => {
        const rawData = snapshot.val();
        const expenses = Object.keys(rawData).reduce((prev, current) => {
          prev.push({ ...rawData[current], id: current });
          return prev;
        }, []);

        return dispatch(setExpenses(expenses));
      });
  };
};
