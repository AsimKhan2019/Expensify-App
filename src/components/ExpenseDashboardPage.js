import React from 'react';

export const ExpenseDashboardPage = () => (
  <p>This is from dashboard Component</p>
);
