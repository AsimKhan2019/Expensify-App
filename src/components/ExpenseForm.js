import React from 'react';
import moment from 'moment';

import { SingleDatePicker } from 'react-dates';
import 'react-dates/lib/css/_datepicker.css';

export default class ExpenseForm extends React.Component {
  constructor(props) {
    super(props);
  }
  defaultState = () => ({
    description: '',
    note: '',
    amount: '',
    createdAt: moment(),
    calenderfocused: false,
    error: undefined
  });

  state = {
    ...this.defaultState(),
    description: this.props.expense ? this.props.expense.description : '',
    note: this.props.expense ? this.props.expense.note : '',
    amount: this.props.expense ? this.props.expense.amount : '',
    createdAt: this.props.expense
      ? moment(this.props.expense.createdAt)
      : moment()
  };

  onDescriptionChange = e => {
    const description = e.target.value;
    this.setState(() => ({ description }));
  };

  onNoteChange = e => {
    const note = e.target.value;
    this.setState(() => ({ note }));
  };

  onAmountChange = e => {
    const amount = e.target.value.trim();
    if (amount && !amount.match(/^\d{1,}(\.\d{0,2}){0,1}$/)) return;
    this.setState(() => ({ amount }));
  };

  onDateChange = createdAt => {
    if (!createdAt) return;
    this.setState(() => ({ createdAt }));
  };

  onCalenderFocusChange = ({ focused }) => {
    this.setState(() => ({ calenderfocused: focused }));
  };

  onSubmitHandler = e => {
    e.preventDefault();
    const description = this.state.description.trim();
    if (!description || !this.state.amount) {
      this.setState(() => ({
        error: 'Description & Amount fields are required'
      }));
      return;
    }
    this.props.onSubmit({
      amount: +this.state.amount,
      createdAt: this.state.createdAt.valueOf(),
      note: this.state.note,
      description: description
    });
    this.setState(() => ({
      ...this.defaultState()
    }));
  };

  render() {
    return (
      <div>
        {this.state.error && <p>{this.state.error}</p>}
        <form onSubmit={this.onSubmitHandler}>
          <input
            value={this.state.description}
            onChange={this.onDescriptionChange}
            type="text"
            placeholder="Description"
            autoFocus
          />
          <input
            value={this.state.amount}
            onChange={this.onAmountChange}
            type="text"
            placeholder="Amount"
          />
          <SingleDatePicker
            date={this.state.createdAt}
            onDateChange={this.onDateChange}
            focused={this.state.calenderfocused}
            onFocusChange={this.onCalenderFocusChange}
            numberOfMonths={1}
            isOutsideRange={() => false}
            displayFormat="DD/MM/YY"
          />
          <textarea
            value={this.state.note}
            onChange={this.onNoteChange}
            placeholder="Add a note for expense (optional)"
          />
          <button>{this.props.submitBtnName || 'Add Expense'}</button>
        </form>
      </div>
    );
  }
}
