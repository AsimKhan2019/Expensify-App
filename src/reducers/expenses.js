const expensesReducerDefaultState = () => [];

export const expensesReducer = (
  state = expensesReducerDefaultState(),
  action
) => {
  switch (action.type) {
    case 'ADD_EXPENSE':
      return [...state, action.expense];
    case 'REMOVE_EXPENSE':
      return state.filter(e => e.id !== action.expense.id);
    case 'EDIT_EXPENSE':
      return state.map(
        e =>
          e.id === action.expense.id
            ? { ...e, ...action.expense, id: action.expense.id }
            : e
      );
    case 'SET_EXPENSE':
      return action.expenses;
    default:
      return state;
  }
};
