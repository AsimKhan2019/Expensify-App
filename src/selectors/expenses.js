import moment from 'moment';

export const getFilteredExpense = (
  expenses = [],
  { endDate, startDate, sortBy = 'date', text = '' } = {}
) => {
  return expenses
    .filter(expense => {
      const createdAtMoment = moment(expense.createdAt);
      const startDateMatch = !startDate
        ? true
        : startDate.isSameOrBefore(createdAtMoment);
      const endDateMatch = !endDate
        ? true
        : endDate.isSameOrAfter(createdAtMoment);

      const textMatch = expense.description
        .trim()
        .toLowerCase()
        .includes(text.trim().toLowerCase());

      return startDateMatch && endDateMatch && textMatch;
    })
    .sort((a, b) => {
      if (sortBy === 'amount') return a.amount < b.amount ? 1 : -1;

      return a.createdAt < b.createdAt ? 1 : -1;
    });
};
