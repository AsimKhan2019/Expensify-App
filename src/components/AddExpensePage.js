import React from 'react';
import ExpenseForm from './ExpenseForm';
import { connect } from 'react-redux';
import { startAddExpense } from '../actions/expenses';

export const AddExpensePage = props => (
  <div>
    <h2>Add Expense</h2>
    <ExpenseForm
      onSubmit={expense => {
        props.startAddExpense(expense);
        props.history.push('/');
      }}
    />
  </div>
);

const mapDispatchToProps = dispatch => ({
  startAddExpense: expense => dispatch(startAddExpense(expense))
});

export default connect(undefined, mapDispatchToProps)(AddExpensePage);
