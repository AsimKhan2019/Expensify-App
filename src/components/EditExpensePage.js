import React from 'react';
import { connect } from 'react-redux';

import ExpenseForm from './ExpenseForm';
import { editExpense } from '../actions/expenses';

export const EditExpensePage = props => (
  <div>
    <h2>Edit Expense</h2>
    <ExpenseForm
      submitBtnName="Update Expense"
      expense={props.expense}
      onSubmit={expense => {
        props.editExpense(props.match.params.id, expense);
        props.history.push('/');
      }}
    />
  </div>
);

const mapStateToProps = (state, props) => {
  const expenseId = props.match.params.id;
  const expense = state.expenses.find(e => e.id === expenseId);
  return { expense };
};

const mapDispatchToProps = dispatch => {
  return {
    editExpense: (id, expense) => {
      dispatch(editExpense(id, expense));
    }
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);
