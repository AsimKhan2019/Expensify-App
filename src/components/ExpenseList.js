import React from 'react';
import { connect } from 'react-redux';

import ExpenseListItem from './ExpenseListItem';
import { getFilteredExpense } from '../selectors/expenses';

export const ExpenseList = props => (
  <div>
    <h1>Expense list</h1>
    {(!props.expenses || props.expenses.length === 0 )? (
      <p>No Expense</p>
    ) : (
      props.expenses.map(e => <ExpenseListItem key={e.id} {...e} />)
    )}
  </div>
);

const mapStateToProps = state => ({
  expenses: getFilteredExpense(state.expenses, state.filters)
});

export default connect(mapStateToProps)(ExpenseList);
