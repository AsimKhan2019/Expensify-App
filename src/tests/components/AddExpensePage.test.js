import React from 'react';
import { shallow } from 'enzyme';

import ExpenseForm from '../../components/ExpenseForm';
import { AddExpensePage } from '../../components/AddExpensePage';

import { expenses } from '../fixtures/expenses';

let addExpense, history, wrapper;

beforeEach(() => {
  addExpense = jest.fn();
  history = { push: jest.fn() };
  wrapper = shallow(<AddExpensePage addExpense={addExpense} history={history} />);
});

test('should render add expense page correctly', () => {
  expect(wrapper).toMatchSnapshot();
});

test('should handle on submit', () => {
  wrapper.find(ExpenseForm).prop('onSubmit')(expenses[0]);

  expect(history.push).toHaveBeenCalled();
  expect(addExpense).toHaveBeenCalled();
});
