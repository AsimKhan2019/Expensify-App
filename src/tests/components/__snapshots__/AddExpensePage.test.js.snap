// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render add expense page correctly 1`] = `
<div>
  <div
    className="page-summary"
  >
    <div
      className="container"
    >
      <h2>
        Add Expense
      </h2>
    </div>
  </div>
  <ExpenseForm
    onSubmit={[Function]}
  />
</div>
`;
